sandbox/lua/decoders/bind_query_log.lua

```diff
@@ -48,197 +45,45 @@ local date_time = require 'date_time'
-
-]]--
-
---Create a table to hold possible matches that may appear:
-local t = {
-      ["+"] = "recursion requested",
-      ["-"] = "recursion not requested",
-      E = "EDNS used",
-      S = "query signed",
-      D = "DNSSEC data wanted",
-      C = "no DNSSEC validation check",
-      T = "TCP used",
+local msg = {
+  --This value is read in from the 'msg_type' config option in the TOML:
+  Type        = msg_type,
+  Payload     = nil,
+  Severity    = 'info',
+  Fields      = {},
 }
 
---Create a capture group that uses the table above to match any one or more of '+-EDCST'
---that are present and add them to a new table called QueryFlags: 
-query_flags = l.Cg( l.Ct((l.S"+-EDCST" / t)^1), "QueryFlags" )
-
---[[Hostname and domain name patterns
-
-Hostnames and domain names are broken up into fragments that are called 
-"labels", which are the parts between the dots.
-
-Source: https://en.wikipedia.org/wiki/Hostname#Restrictions_on_valid_host_names
-
-For instance, webserver.company.com has the labels "webserver", "company" and 
-"com"
-
-The pattern below uses the upper, lower and digit shortcuts from the main LPEG 
-library and combines them with the hyphen (-) character to match anything that's
-part of a valid hostname label. The ^1 means match one or more instances of it.
---]]
-local hostname_fragment = (l.upper + l.lower + l.digit +  "-")^1
+function process_message ()
 
---The pattern below matches one or more hostname_fragments, followed by a . 
---and followed by one more hostname_fragment, indicating the end of a complete
---hostname. The open and close parens and colon are to match the decorations
---BIND puts around the name: 
--- (webserver.company.com):
-local enclosed_query = "(" * l.Cg((hostname_fragment * ".")^1 * hostname_fragment, "FullQuery") * "):"
+  --Create a local variable for the log line from the Payload of the incoming Heka message.
+  --The LogstreamerInput that generates the messages automatically puts the 
+  local query_log_line = read_message("Payload")
+  --Use the bind_query grammar defined farther above to create a local grammar object:
+  local grammar = l.Ct(bind.query_log_grammar)
+  --Create a fields table and use the :match method on the grammar object to fill it with the Lua table
+  --of values generated by parsing the query_log_line:
+  fields = grammar:match(query_log_line)
 
---The ^-1 means match at most 1 instance of the pattern. We want this so that we 
---can match the first part of a hostname and leave the rest for the l.Cg((hostname_fragment...
---capture group to match into the QueryDomain table entry.
---In webserver.company.com, `(hostname_fragment * ".")^-1` matches webserver.
---and `l.Cg((hostname_fragment...` matches company.com
-local query = l.Cg((hostname_fragment)^-1, "QueryName") * "." * l.Cg((hostname_fragment * ".")^1 * hostname_fragment, "QueryDomain")
+  --If fields is empty, exit immediately:
+  if not fields then return -1 end
```

Comment:
No, I'm not talking about the `error()` function, I'm talking about the `return -1`. You can return multiple values, and if you return a string for the second value it will be used as the error message. There are a couple of examples in the [json](https://github.com/mozilla-services/heka/blob/be1420d226808485a557ad622bb75785cadfa43a/sandbox/lua/decoders/json.lua#L175) [decoder](https://github.com/mozilla-services/heka/blob/be1420d226808485a557ad622bb75785cadfa43a/sandbox/lua/decoders/json.lua#L201).